os_keystone_endpoint.py

albertomurillo opened the pull request:
##### SUMMARY
This patch adds a new Ansible module which allows a user to create
an endpoint to a service with Keystone

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
os_keystone_endpoint.py

##### ANSIBLE VERSION
```
ansible 2.3.2.0
  config file = /ansible-openstack/test/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]
```


##### ADDITIONAL INFORMATION
At the moment, there is no way to configure specific endpoints for Keystone services. This module allows a user to make all of those changes which can make configuring new OpenStack deployments easier! This pretty much compliments the os_keystone_service module.

This PR is a fork of https://github.com/ansible/ansible/pull/20314 since the original author hasn't responded to feedback.

mnaser committed:
Add os_keystone_service_endpoint

This patch adds a new Ansible module which allows a user to create
an endpoint to a service with Keystone.

Fixes #23909

albertomurillo committed:
os_keystone_endpoint: Fix style and messages

Fix comments, pep8, version, metadata, license header
and imports according to the Contributing Modules Checklist

Signed-off-by: Alberto Murillo <albertomurillosilva@gmail.com>

albertomurillo committed:
os_keystone_endpoint: Fix return values

- Change type of 'endpoint' return value from dictionary to complex
  in order to get validate_module checks passed.

- Remove 'id' from the return data since it is included inside the
  'endpoint' value wich is already being returned.

- Rename 'service' field to 'service_id' which is the correct name
  for the service id field returned in json.

Signed-off-by: Alberto Murillo <albertomurillosilva@gmail.com>

albertomurillo committed:
os_keystone_endpoint: Update shade version

Update minimum shade version to 1.11.0

Signed-off-by: Alberto Murillo <albertomurillosilva@gmail.com>

mgale commented on lib/ansible/modules/cloud/openstack/os_keystone_endpoint.py:
```diff
@@ -0,0 +1,213 @@
+     default: present
+requirements:
+    - "python >= 2.6"
+    - "shade"
```
If requirement versions are specified here does that get used by a build system anywhere or is this purely a docstring for the user?

albertomurillo replied:
Good question! I think `python >= 2.6` can be removed since the developing new modules page states:
```
Adding Python 3 support to modules (all new modules must be Python-2.6 and Python-3.5 compatible).
```

albertomurillo replied:
I guess this is just docstring for the user

mgale reviewed (commented):
I add a comment, mainly for my own understanding.

SamYaple commented on lib/ansible/modules/cloud/openstack/os_keystone_endpoint.py:
```diff
@@ -0,0 +1,213 @@
+            description: Region Name.
+            type: string
+            sample: "RegionOne"
+        service_id:
```
this isn't correct according to the code below, should be 'service' and accept name_or_id

albertomurillo replied:
`service_id` is actually correct, this is the info returned.

```
TASK [debug] **********************************************************************************************************************************************************************
ok: [controller.example.com] => {
    "msg": {
        "changed": true,
        "msg": "All items completed",
        "results": [
            {
                "_ansible_item_result": true,
                "_ansible_no_log": false,
                "_ansible_parsed": true,
                "changed": true,
                "endpoint": {
                    "HUMAN_ID": false,
                    "NAME_ATTR": "name",
                    "enabled": true,
                    "human_id": null,
                    "id": "34ab50cb74834ec8911709220eaa92af",
                    "interface": "admin",
                    "links": {
                        "self": "http://controller.example.com:5000/v3/endpoints/34ab50cb74834ec8911709220eaa92af"
                    },
                    "region": "RegionOne",
                    "region_id": "RegionOne",
                    "service_id": "4e26c3decfdb4d37907fc798b2828fd4",
                    "url": "http://controller.example.com:9292"
                },
                "invocation": {
                    "module_args": {
                        "api_timeout": null,
                        "auth": null,
                        "auth_type": null,
                        "availability_zone": null,
                        "cacert": null,
                        "cert": null,
                        "cloud": "default",
                        "enabled": true,
                        "endpoint_type": "public",
                        "interface": "admin",
                        "key": null,
                        "region": "RegionOne",
                        "region_name": null,
                        "service": "glance",
                        "state": "present",
                        "timeout": 180,
                        "url": "http://controller.example.com:9292",
                        "verify": null,
                        "wait": true
                    }
                },
                "item": [
                    {
                        "service": "glance"
                    },
                    {
                        "interface": "admin",
                        "url": "http://controller.example.com:9292"
                    }
                ]
            }
        ]
    }
}
```

albertomurillo replied:
I think you mean in line 23 should accept service's name or id as input.

albertomurillo replied:
Fixed

SamYaple commented on lib/ansible/modules/cloud/openstack/os_keystone_endpoint.py:
```diff
@@ -0,0 +1,213 @@
+        service=dict(required=True),
+        interface=dict(required=True, choices=['admin', 'public', 'internal']),
+        url=dict(required=True),
+        region=dict(required=True),
```
region should not be required and hsould default to None

albertomurillo replied:
Fixed

SamYaple commented on lib/ansible/modules/cloud/openstack/os_keystone_endpoint.py:
```diff
@@ -0,0 +1,213 @@
+    try:
+        cloud = shade.operator_cloud(**module.params)
+
+        service = cloud.get_service(service_name)
```
service could be None in this case, we should be checking that before we use 'service.id' in the next line

albertomurillo replied:
Fixed

SamYaple commented on lib/ansible/modules/cloud/openstack/os_keystone_endpoint.py:
```diff
@@ -0,0 +1,213 @@
+
+        if state == 'present':
+            if endpoint is None:
+                result = cloud.create_endpoint(service_name_or_id=service.id,
```
if you pass the full object in rather than the service.id you can save a lookup from the extra GET, but you would need to bump up the minimum version of shade a bit

albertomurillo replied:
will check this one in a moment

albertomurillo replied:
I don't understand how to pass the full object.

```python
    @_utils.valid_kwargs('public_url', 'internal_url', 'admin_url')
    def create_endpoint(self, service_name_or_id, url=None, interface=None,
                        region=None, enabled=True, **kwargs):
```

SamYaple replied:
result = cloud.create_endpoint(service_name_or_id=service,

albertomurillo replied:
I understand now and fixed. Why not make that ["feature"](https://github.com/openstack-infra/shade/commit/0d5e7b561ea4e0404157cabbcd1e6b56d633c079) documented?

SamYaple replied:
Thats a question for @emonty

albertomurillo committed:
os_keystone_endpoint: Make region optional

Signed-off-by: Alberto Murillo <albertomurillosilva@gmail.com>

albertomurillo committed:
os_keystone_endpoint: Validate service exists before using service.id

Signed-off-by: Alberto Murillo <albertomurillosilva@gmail.com>

albertomurillo committed:
os_keystone_endpoint: Fix documentation for service to accept name or id

Signed-off-by: Alberto Murillo <albertomurillosilva@gmail.com>

albertomurillo committed:
os_keystone_endpoint: Pass the full service object to create_endpoint()

We already have the service object retrieved in code, by passing service.id to
create_endpoint, the shade librarie queries the api again to get the full service
object.

By Passing the already rerieved service object to create_endpoint() we save one
request to the API.

Signed-off-by: Alberto Murillo <albertomurillosilva@gmail.com>

SamYaple commented on lib/ansible/modules/cloud/openstack/os_keystone_endpoint.py:
```diff
@@ -0,0 +1,217 @@
+  os_keystone_endpoint:
+     cloud: mycloud
+     state: absent
+     name: image
```
should this be 'service: image'

albertomurillo replied:
its actually service name or id so `service: glance`

albertomurillo replied:
fixed

SamYaple commented on lib/ansible/modules/cloud/openstack/os_keystone_endpoint.py:
```diff
@@ -0,0 +1,217 @@
+            if endpoint is None:
+                result = cloud.create_endpoint(service_name_or_id=service,
+                                               url=url, interface=interface,
+                                               region=region, enabled=True)
```
should be passing enabled=enabled

albertomurillo replied:
fixed

SamYaple commented on lib/ansible/modules/cloud/openstack/os_keystone_endpoint.py:
```diff
@@ -0,0 +1,217 @@
+            else:
+                if _needs_update(module, endpoint):
+                    endpoint = cloud.update_endpoint(
+                        endpoint.id, url=url, enabled=enabled)
```
service_name_or_id, interface, and region should also be passed here since that is all updateable

albertomurillo replied:
not quite.

This module does not accept the endpoint id, it takes service_name_or_id, interface and region to find the endpoint. If the user specifies another name_or_id, interface or region then a matching endpoint will not be found and a new one will be created.

SamYaple replied:
ah i see your point. yea that makes sense. well this looks good to me then. thanks for the changes!

ghost commented:
shipit

dagwieers commented on lib/ansible/modules/cloud/openstack/os_keystone_endpoint.py:
```diff
@@ -0,0 +1,217 @@
+#!/usr/bin/python
+# Copyright (c) 2017 VEXXHOST, Inc.
```
The syntax for copyright line is:
```python
# Copyright: (c) 2017, VEXXHOST, Inc.
```
And I'd add a blank line before.

albertomurillo replied:
The [dev guide](http://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#copyright) says otherwise

```
#!/usr/bin/python
# Copyright (c) 2017 Ansible Project
# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
```

dagwieers replied:
Yes, the dev guide did something different than a major rewrite did. I'm trying to get clarification on this.

dagwieers replied:
Ok, so we updated the dev guide to reflect the rewrite.
Now, this is not a deal-breaker, if you don't like or want it, I don't care :-)

albertomurillo replied:
fixed and kudos for fixing the doc.

dagwieers commented on lib/ansible/modules/cloud/openstack/os_keystone_endpoint.py:
```diff
@@ -0,0 +1,217 @@
+     choices: [present, absent]
+     default: present
+requirements:
+    - "shade>=1.11.0"
```
No need to quote this line. Please add spaces around `>=`

albertomurillo replied:
Fixed

dagwieers commented on lib/ansible/modules/cloud/openstack/os_keystone_endpoint.py:
```diff
@@ -0,0 +1,217 @@
+        region:
+            description: Region Name.
+            type: string
+            sample: "RegionOne"
```
No need to quote this string.

albertomurillo replied:
Fixed

dagwieers commented on lib/ansible/modules/cloud/openstack/os_keystone_endpoint.py:
```diff
@@ -0,0 +1,217 @@
+        service_id:
+            description: Service ID.
+            type: string
+            sample: "b91f1318f735494a825a55388ee118f3"
```
No need to quote string.

albertomurillo replied:
Fixed

dagwieers commented on lib/ansible/modules/cloud/openstack/os_keystone_endpoint.py:
```diff
@@ -0,0 +1,217 @@
+        interface:
+            description: Endpoint Interface.
+            type: string
+            sample: "public"
```
No need to quote.

albertomurillo replied:
Fixed

dagwieers commented on lib/ansible/modules/cloud/openstack/os_keystone_endpoint.py:
```diff
@@ -0,0 +1,217 @@
+        url:
+            description: Service URL.
+            type: string
+            sample: "http://controller:9292"
```
No need to quote.

albertomurillo replied:
Fixed

dagwieers commented on lib/ansible/modules/cloud/openstack/os_keystone_endpoint.py:
```diff
@@ -0,0 +1,217 @@
+
+def main():
+    argument_spec = openstack_full_argument_spec(
+        service=dict(required=True),
```
I'd prefer having the type specified for all parameters, like this:
```python
   argument_spec = openstack_full_argument_spec(
        service=dict(type='str', required=True),
        interface=dict(type='str', required=True, choices=['admin', 'internal', 'public']),
        url=dict(type='str', required=True),
        region=dict(type='str'),
        enabled=dict(type='bool', default=True),
        state=dict(type='str', default='present', choices=['absent', 'present']),
    )
```

albertomurillo replied:
Fixed

albertomurillo committed:
os_keystone_endpoint: Make type explicit in module arguments.

Althoug type is default to str when not specified in module arguments
this commit explicitly defines type='str' for better readability.

Signed-off-by: Alberto Murillo <albertomurillosilva@gmail.com>

dagwieers commented on lib/ansible/modules/cloud/openstack/os_keystone_endpoint.py:
```diff
@@ -0,0 +1,217 @@
+        service=dict(type='str', required=True),
+        interface=dict(type='str', required=True, choices=['admin', 'public', 'internal']),
+        url=dict(type='str', required=True),
+        region=dict(type='str', required=False, default=None),
```
So we tend not to specify `required=False` and `default=None` because that is the case by default.
(Although we still like to state the type, even the default nowadays is "str")

Some of these rules are codified, I would like the sanity tests to enforce some of this.

dagwieers replied:
So the main reason for this is that it increases readability.
If only the items that are required are listed like this, you can see with a glance which parameters are required.
Same for default values. It's useful to easily identify the custom default values.
(Just like PEP8, this is purely from a readability point-of-view)

For the same reason we prefer lists to be alphabetically sorted, unless the order has a purpose.

albertomurillo replied:
fixed

dagwieers reviewed (approved):
LGTM

dagwieers commented:
So the nice thing is that you only need 2 shipits, and I'll leave that up to the Ansible OpenStack community ;-)

s-hertel commented:
bot_status

ghost commented:
I do not understand the ansibot very well, but I do know that I have not requested any changes and am happy to see this merged as is. Some one please let me know where ive messed up

deckTECHeu commented:
Hello,
I still do not see ansible module: os_keystone_endpoint.
Any ideas when it will be created?

ghost commented:
@deckTECHeu its merged https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/openstack/os_keystone_endpoint.py

It will be in ansible 2.5

s-hertel commented:
@deckTECHeu If you need the module now, you can copy it into your local `library/`.
